### PR TITLE
Avoid IsValueType in ApplyNullableTransforms

### PR DESCRIPTION
**Pull request by cston**
Avoid calling `IsValueType` in `ApplyNullableTransforms` since `ApplyNullableTransforms` may be called while the `TypeSymbol` is being constructed. (In the case of #29041, when constructing a `PETypeParameterSymbol` where the constraints reference the type parameter.)

Calling `IsValueType` in `ApplyNullableTransforms` is not necessary since the `IsAnnotated` bit is only set in the `[Nullable(new bool[])]` attribute for nullable reference types.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs [hunk 1]
```diff
@@ -654,17 +654,9 @@ public bool ApplyNullableTransforms(ImmutableArray<bool> transforms, INonNullTyp
...
-                {
-                    result = result.AsNullableReferenceType();
-                }
-                else
-                {
-                    result = result.AsNotNullableReferenceType();
-                }
-            }
+            result = isAnnotated ?
+                result.AsNullableReferenceType() :
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs]: Not related to this PR: we should probably rename these methods (`WithIsAnnotated`?) at some point.

**Comment by cston** [comment]
@dotnet/roslyn-compiler, please review.
@jcouv, the PR now includes a fix for #29048. Please take a second look, thanks.

File: src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs [hunk 2]
```diff
@@ -709,7 +709,11 @@ public override bool Equals(object other)
         public override int GetHashCode()
 #pragma warning restore CS0809
         {
-            throw ExceptionUtilities.Unreachable;
+            if (_defaultType is null)
+            {
+                return 0;
+            }
+            return Hash.Combine(_defaultType.GetHashCode(), IsAnnotated.GetHashCode());
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs]: Should we also include `_extensions` in the mix?

File: src/Compilers/CSharp/Portable/Emitter/Model/NamedTypeSymbolAdapter.cs [hunk 3]
```diff
@@ -467,7 +467,7 @@ IEnumerable<Cci.TypeReferenceWithAttributes> Cci.ITypeDefinition.Interfaces(Emit
                 var type = TypeSymbolWithAnnotations.Create(@interface, isNullableIfReferenceType: null);
                 yield return type.GetTypeRefWithAttributes(
                     moduleBeingBuilt,
-                    this.DeclaringCompilation,
+                    this,
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Emitter/Model/NamedTypeSymbolAdapter.cs]: >this [](start = 20, length = 4)

nit: consider naming argument

File: src/Compilers/CSharp/Portable/Emitter/Model/TypeParameterSymbolAdapter.cs [hunk 4]
```diff
@@ -259,7 +259,7 @@ IEnumerable<Cci.TypeReferenceWithAttributes> Cci.IGenericParameter.GetConstraint
 
                 yield return type.GetTypeRefWithAttributes(
                                                             moduleBeingBuilt,
-                                                            this.DeclaringCompilation,
+                                                            this,
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Emitter/Model/TypeParameterSymbolAdapter.cs]: >this [](start = 60, length = 4)

nit: here too

File: src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs (hunk 2 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs]: Changed to use `TypeSymbol` and `IsAnnotated` to ensure `GetHashCode()` returns the same value for two instances that would compare `Equals()`.

---
In reply to: [207666724](https://github.com/dotnet/roslyn/pull/29045#discussion_r207666724) [](ancestors = 207666724)

File: src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs (hunk 1 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs]: Added comment.

---
In reply to: [207445706](https://github.com/dotnet/roslyn/pull/29045#discussion_r207445706) [](ancestors = 207445706)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 5]
```diff
@@ -38101,5 +38102,120 @@ void M()
...
+public class A3<T> where T : struct, IEquatable<T> { }
+public class A4<T> where T : struct, IEquatable<T?> { }
+public class A5<T> where T : IEquatable<string?> { }
+public class A6<T> where T : IEquatable<int?> { }";
+            var source =
+@"class B
+{
+    static void Main()
+    {
+        new A0<string?>();
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >string? [](start = 15, length = 7)

From our discussion, I expect a warning on `A0<string?>` once the substitution rule is adjusted for oblivious/nullable mixing together.

Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Added comment.

---
In reply to: [207677078](https://github.com/dotnet/roslyn/pull/29045#discussion_r207677078) [](ancestors = 207677078)

**Pull request merged**